standardized-metadata-flexible-spec.py

```diff
@@ -0,0 +1,219 @@
+
+    num_rows_total: Optional[int]
+    sample: Optional[Sequence[TableRecord]]
+
+
+class TableMaterializationMetadata(TableObservationMetadata):
+    """Metadata fields that apply to materializations of assets that are tables."""
+
+    num_rows_inserted: Optional[int]
+    num_rows_updated: Optional[int]
+    num_rows_deleted: Optional[int]
+
+
+####################################################################################################
+# Metadata that's relevant to assets that are Snowflake tables. Lives inside the "dagster_snowflake"
+# package.
+####################################################################################################
+
+
+class SnowflakeTableAddress(BaseModel):
+    """An identifier for a table in Snowflake."""
+
+    database: Optional[str]
+    db_schema: Optional[str]
+    table_name: Optional[str]
+
+
+class SnowflakeTableMetadata(NamespacedMetadata):
+    """Metadata fields that apply to assets that are tables in Snowflake."""
+
+    snowflake_address: Optional[SnowflakeTableAddress]
+    cluster_by: Optional[str]
+
+    @classmethod
+    def namespace(cls) -> str:
+        return "dagster_snowflake"
+
+
+####################################################################################################
+# An asset with metadata constructed using the structured metadata APIs.
+####################################################################################################
+
+
+@asset
+def asset1():
+    return MaterializeResult(
+        metadata={
+            **AssetMetadata(
+                storage_kind="snowflake", storage_address_string="my_db.my_schema.asset1"
+            ),
+            **TableMaterializationMetadata(
+                num_rows_total=500,
+                num_rows_inserted=5,
+                columns=[TableColumn(name="user_id", data_type="str")],
+            ),
+            **SnowflakeTableMetadata(
+                snowflake_address=SnowflakeTableAddress(
+                    database="my_db", db_schema="my_schema", table_name="asset1"
+                )
+            ),
+            # non-structured metadata
+            "my_unschematized_piece_of_metadata": 5,
+            "staging_address": SnowflakeTableAddress(
+                database="my_staging_db", db_schema="my_schema", table_name="asset1"
+            ).dict(),
+        }
+    )
+
+
+####################################################################################################
+# An asset with metadata constructed using plain JSON-objects. Has identical metadata to the asset
+# above.
+####################################################################################################
+
+
+@asset
+def asset2():
+    return MaterializeResult(
+        metadata={
+            "dagster.table/num_rows_inserted": 5,
+            "dagster.table/columns": [{"name": "user_id", "data_type": "str"}],
+            "dagster.table/num_rows_total": 500,
```

Comment:
what happens if i try to emit metadata here that isn't part of the class? e.g. dagster.table/name: foo

Reply:
We can't stop you from doing this (in either of the variants). However, you could run something like `TableMaterializationMetadata.validate(metadata)` to assert that you're yielding known entries.

Reply:
But the metadata will still make it through? I think that's fine, better than an error.

Reply:
Right, the metadata will still make it through